language/src/regent/optimize_index_launches.t

```diff
@@ -1219,140 +1248,242 @@ local function init_bitmask(bitmask, volume)
+                 util.mk_expr_binary("*",
+                   util.mk_expr_field_access(pf, dim_names[dim], int64),
+                   extent),
+                 prev_ast)
+    return recurse(dim - 1, extent, ast)
+  end
 
-      return util.mk_expr_binary("+",
-        util.mk_expr_binary("*", x, util.mk_expr_binary("*", y_extent, z_extent)),
-        util.mk_expr_binary("+", util.mk_expr_binary("*", y, z_extent), z))
+  if dim <= 1 then
+    return pf
+  else
+    return recurse(
+      dim - 1,
+      util.mk_expr_constant(1, int64),
+      util.mk_expr_field_access(pf, dim_names[dim], int64))
+  end
+end
 
+local function collect_and_sort_args(args, call_args, task, param_region_types)
+  -- Collect all args of the same partition together
+  local collected = {}
+  for _, arg in pairs(args) do
+    if type(arg) == "number" then
+      local psym = get_partition_symbol(call_args[arg])
+      collected[psym] = collected[psym] or {}
+      collected[psym][arg] = true
     else
-      assert(false)
+      -- Cross-check args are pairs
+      local psym = get_partition_symbol(call_args[arg[1]])
+      collected[psym] = collected[psym] or {}
+      collected[psym][arg[1]] = true
+      collected[psym][arg[2]] = true
     end
   end
-end
 
-local function insert_dynamic_check(is_demand, args_need_dynamic_check, index_launch_ast, unopt_loop_ast)
-  local check = terralib.newlist()
+  local array = {}
+  for psym, args in pairs(collected) do
+    array[psym] = {}
+    for arg, v in pairs(args) do
+      if v then table.insert(array[psym], arg) end
+    end 
+  end
 
-  local index_types = { [0] = std.int1d, std.int1d, std.int2d, std.int3d }
-  local rect_types = { [0] = std.rect1d, std.rect1d, std.rect2d, std.rect3d }
+  for _, args in pairs(array) do
+    table.sort(args, function(p, q)
+      local p_privileges, _, _, _, _ = std.find_task_privileges(param_region_types[p], task)
+      local q_privileges, _, _, _, _ = std.find_task_privileges(param_region_types[q], task)
 
-  local verdict = std.newsymbol(bool, "verdict")
-  check:insert(util.mk_stat_var(verdict, bool, util.mk_expr_constant(false, bool)))
+      local p_has_WR = p_privileges[1] == "reads_writes" or string.find(p_privileges[1], "reduces")
+      local q_has_WR = q_privileges[1] == "reads_writes" or string.find(q_privileges[1], "reduces")
 
-  for _, idx in pairs(args_need_dynamic_check) do
-    local arg = index_launch_ast.call.args[idx]
-    local stats = terralib.newlist()
+      if p_has_WR and not q_has_WR then
+        return true
+      elseif q_has_WR and not p_has_WR then
+        return false
+      end
+      return p < q
+    end)
+  end
 
-    -- Skip non-partition args
-    if arg:is(ast.typed.expr.IndexAccess) and
-       std.is_region(arg.expr_type) and
-       std.is_partition(std.as_read(arg.value.expr_type)) then
+  return array
+end
 
-      local dim = arg.expr_type:ispace().dim
+local c = terralib.includecstring([[
+#include <stdio.h>
+#include <stdlib.h>
+]])
```

Comment:
Instead of including C headers in the middle of this file, let's use `std.c`, which I think already includes a superset of these headers.

Reply:
Done.